### PR TITLE
Addition of a custom `say-serializer`

### PR DESCRIPTION
**Pull request by elpoelma**
### Overview
This PR adds a new `SaySerializer` class, which is a custom implementation of the default ProseMirror `DOMSerializer`. `SaySerializer` takes the new `serialize` method (which can be defined on node- and mark-specs) into account. If a node- or mark-spec has a `serialize` method defined, it takes priority over the `toDOM` method. An instance of `SaySerializer` is run each time the html-content of the editor is requested, and when the content is copied to the clipboard.

This PR introduces also introduces two new interfaces to support the new `SaySerializer`:
- `SayNodeSpec`: a custom `NodeSpec` which supports the `serialize` method
- `SayMarkSpec`: a custom `MarkSpec` which supports the `serialize` method

By default, the editor uses a instance of the new `SaySerializer`.

#### Differences between `toDOM` and `serialize`
Whereas `toDOM` is called each time a node needs to be converted to its corresponding node-view, `serialize` is only called when an instance of `SaySerializer` is executed. 
On top of the arguments the `toDOM` method receives, `serialize` gets an additional `EditorState` argument.

#### Why do we need this?
The `serialize` method can be useful for editor-nodes which need access to the current editor-state when serializing to a DOM representation. E.g. having access to the `ember-intl` service, or having access to the whole editor document can be valid usecases.

### How to test/reproduce
The sample `counter` ember-node has been updated to support the new `serialize` method. Depending on the document language, it's serialized version will either be translated in dutch or english.

### Checks PR readiness
- [x] changelog
- [x] npm lint
- [x] no new deprecations

### CODE REVIEW

**Review by x-m-el — CHANGES_REQUESTED** [review]
`Whereas toDOM is called each time a node needs to be converted to its corresponding node-view, serialize is only called when an instance of SaySerializer is executed.`
I'm not sure what this means/what the impact of this is.
Does this mean that in some instances both a `toDOM` and `serialize` function can be useful, as they fire at different times? Or could you (in theory) just always use `serialize` from now on, as it is the same as `toDOM` but with some extra input you can use?


I did test it and looked at the code.
Everything seems to work like it should and It looks like a good way to do it.
I'm not 100% clear on how SaySerializer class works (as it is quite low-level), so if you do need extra eyes on that, you might want to wait for a review of @abeforgit.



"unrelated" note:
A small issue is that we use all these properties through the EmberNodeConfig type interface and getting to the actual source (and explanation/details what it does) is hard (or at least I couldn't with ctrl+clicking :p )
Like for `serialize` here and for serialize/parse in attrs.
Might need to document this specific place a bit better to make it easier to use these ember-nodes + reduce potential wrong assumptions we could make.
Because I assume I know what serialize and parse for an attribute does, but I can't be sure.

File: addon/core/say-serializer.ts
Comment (elpoelma) [reply addon/core/say-serializer.ts:140]: Alright, I updated the comments to clarify the `serialize` method takes precendence.

**Comment by elpoelma** [comment]
> `Whereas toDOM is called each time a node needs to be converted to its corresponding node-view, serialize is only called when an instance of SaySerializer is executed.` I'm not sure what this means/what the impact of this is. Does this mean that in some instances both a `toDOM` and `serialize` function can be useful, as they fire at different times? Or could you (in theory) just always use `serialize` from now on, as it is the same as `toDOM` but with some extra input you can use?
> 
> I did test it and looked at the code. Everything seems to work like it should and It looks like a good way to do it. I'm not 100% clear on how SaySerializer class works (as it is quite low-level), so if you do need extra eyes on that, you might want to wait for a review of @abeforgit.
> 
> "unrelated" note: A small issue is that we use all these properties through the EmberNodeConfig type interface and getting to the actual source (and explanation/details what it does) is hard (or at least I couldn't with ctrl+clicking :p ) Like for `serialize` here and for serialize/parse in attrs. Might need to document this specific place a bit better to make it easier to use these ember-nodes + reduce potential wrong assumptions we could make. Because I assume I know what serialize and parse for an attribute does, but I can't be sure.

In some instances both a `toDOM` and `serialize` method can indeed be useful. `serialize` is only used by the an instance of `SaySerializer` and is not used to construct node-views. `toDOM` is used by the `SaySerializer` if `serialize` is not defined, but is also used to construct node-views.

On the ember-nodes: I agree that we should improve the types for them, unsure whether to do this in this PR or in a seperate one.

File: addon/core/say-mark-spec.ts
Comment (x-m-el) [inline addon/core/say-mark-spec.ts:14]:
```suggestion
   * Note: this method is not a replacement to `toDOM`, `toDOM` is still used by Prosemirror to construct the node-view for this node,
   * `serialize` is only used when `SaySerializer` is explicitly used in our code to serialize something.
   * e.g. copying a part of the editor (see `clipboardSerializer`) or extracting the document html using the 'SayController.htmlContent` method.
```
- small typo
- I tried to write it a bit more explicit. I'm not 100% convinced it's better, but I think this makes it more clear that this is just a function that we might use ourselves in places, not that something gets "override" in Prosemirror.

**Review by x-m-el — APPROVED** [review]
> In some instances both a `toDOM` and `serialize` method can indeed be useful. `serialize` is only used by the an instance of `SaySerializer` and is not used to construct node-views. `toDOM` is used by the `SaySerializer` if `serialize` is not defined, but is also used to construct node-views.
> 
> On the ember-nodes: I agree that we should improve the types for them, unsure whether to do this in this PR or in a seperate one.

1. Alright,I get it now (I got confused somewhere which tripped me up completely). It's a lot more simple than I first thought. Which is good, less magic means less confusion and mistakes :) 

2. Definitely separate ticket. This review was the reason I noticed it, so I wrote it down here to not forget + already get a second opinion. If you agree this is needed, making a ticket for it would be perfect :)
Like I said, ember-nodes are quite understandable, but documentation will give a certainty how something works. Also, there is a mix of custom functions and functions directly from Prosemirror. Somehow making clear which is which can also be useful (as for one you should just check out Prosemirror documentation). Obviously it is not sensible to "copy" over Prosemirror docs for those specific parameters.

3. Small comment about changing some wording. Feel free to decide if you agree. No other comments regarding the code.

**Comment by elpoelma** [comment]
> > In some instances both a `toDOM` and `serialize` method can indeed be useful. `serialize` is only used by the an instance of `SaySerializer` and is not used to construct node-views. `toDOM` is used by the `SaySerializer` if `serialize` is not defined, but is also used to construct node-views.
> > On the ember-nodes: I agree that we should improve the types for them, unsure whether to do this in this PR or in a seperate one.
> 
>     1. Alright,I get it now (I got confused somewhere which tripped me up completely). It's a lot more simple than I first thought. Which is good, less magic means less confusion and mistakes :)
> 
>     2. Definitely separate ticket. This review was the reason I noticed it, so I wrote it down here to not forget + already get a second opinion. If you agree this is needed, making a ticket for it would be perfect :)
>        Like I said, ember-nodes are quite understandable, but documentation will give a certainty how something works. Also, there is a mix of custom functions and functions directly from Prosemirror. Somehow making clear which is which can also be useful (as for one you should just check out Prosemirror documentation). Obviously it is not sensible to "copy" over Prosemirror docs for those specific parameters.
> 
>     3. Small comment about changing some wording. Feel free to decide if you agree. No other comments regarding the code.

2. I'll make a seperate ticket for it, after merging this one :)

3. Agree with the changing in wording, will update :)

**Pull request merged**
